packages/react-next/src/components/Pivot/useOverflow.ts

```diff
@@ -0,0 +1,158 @@
+    for (let i = overflowIndex; i < items.length; i++) {
+      const item = items[i];
+      if (item instanceof HTMLElement && item !== menuButton) {
+        callbacks.setItemIsOverflowing(item, false);
+      }
+    }
+
+    overflowIndex = undefined;
+  };
+
+  // Hides any items that overflow the container, and shows the overflow menu if applicable
+  const hideOverflowItems = (menuButton: HTMLElement, win: Window) => {
+    const container = menuButton.parentElement;
+    if (!container) {
+      return;
+    }
+
+    const previousOverflowIndex = overflowIndex;
+
+    resetOverflowItems(menuButton);
+
+    const { rtl, containerPadding, menuMargin } = getCachedStyle(container, menuButton, win);
+    const containerRect = container.getBoundingClientRect();
+    const containerEnd = (rtl ? -containerRect.left : containerRect.right) - containerPadding;
+    const items = container.children;
+    let lastVisibleElement: HTMLElement | undefined = undefined;
+
+    // Iterate over the items in reverse order, hiding them as necessary until everything fits within the container
+    for (let i = items.length - 1; i >= 0; i--) {
+      const item = items[i];
+      if (!(item instanceof HTMLElement) || item === menuButton) {
+        continue;
+      }
+
+      // Keep track of the last visible element, even if it's pinned
+      if (lastVisibleElement === undefined) {
+        lastVisibleElement = item;
+      }
+
+      // Don't attempt to hide pinned items
+      if (callbacks.isPinned && callbacks.isPinned(item)) {
+        continue;
+      }
+
+      const lastElementRect = lastVisibleElement.getBoundingClientRect();
+      const lastElementEnd = (rtl ? -lastElementRect.left : lastElementRect.right) + menuMargin;
+
+      if (lastElementEnd <= containerEnd) {
+        break; // Everything fits; we're done
+      }
+
+      // Move the item to the overflow
+      callbacks.setItemIsOverflowing(item, true);
+      lastVisibleElement = menuButton;
+      overflowIndex = i;
+    }
+
+    if (overflowIndex !== previousOverflowIndex && callbacks.onOverflowItemsChanged) {
+      callbacks.onOverflowItemsChanged(overflowIndex);
+    }
+  };
+
+  const async = useAsync();
+
+  // Call hideOverflowItems on the first animation frame, and then every time the window resizes
+  React.useLayoutEffect(() => {
+    const menuButton = overflowMenuButtonRef.current;
+    const win = getWindow(menuButton);
+
+    if (menuButton && win) {
+      const updateOverflow = () => hideOverflowItems(menuButton, win);
+      const updateOverflowDebounced = async.debounce(updateOverflow, 16, { leading: true });
+
+      const requestId = win.requestAnimationFrame(updateOverflow);
+      win.addEventListener('resize', updateOverflowDebounced, false);
+
+      return () => {
+        win.cancelAnimationFrame(requestId);
+        win.removeEventListener('resize', updateOverflowDebounced, false);
+        resetOverflowItems(menuButton);
+      };
+    }
+  });

```

Comment:
i dont think we need to run this hook on every render?
can't we have deps `[overflowMenuButtonRef.current]` here?

Reply:
I'm looking into how to capture the dependencies. It's more than just overflowMenuButtonRef.current; it depends on at least:
* Which tab is selected (the selected tab never goes into the overflow)
* Whether the HTMLElements of any of the tabs has been re-rendered by React (which would reset the dynamically-set properties about whether it's in the overflow)
* The width of each tab item, which generally won't change, but technically depends on the header text and any styling of the tab header itself.

Reply:
I looked into it some more, and it turns out that useEffect/useLayoutEffect don't work properly if you use `[ref.current]` as the dependency. It uses the ref's value from the _previous_ render, which ends up not re-running the effect when it should be.
The proper solution is to do all this work in a ref setter function instead of useLayoutEffect, so it gets triggered whenever the ref itself changes. That is complicated by the fact that we have to "listen" for multiple refs (the overflowMenu as well as any of the individual tab items), and coalesce all of those ref updates into a single recalculation after they've all been set. It's all doable but requires a pretty significant change to how it works now. I'm not sure I'll get to it.